plutus-core/plutus-core/src/PlutusCore/Flat.hs

```diff
@@ -404,16 +403,23 @@ deriving newtype instance Flat (Binder TyName)
 deriving newtype instance Flat (Binder NamedDeBruijn)
 deriving newtype instance Flat (Binder NamedTyDeBruijn)
 
--- this instance is very similar to the Flat DeBruijn instance.
--- NOTE: the serialization roundtrip holds iff the invariant name==fakeName holds
+{- This instance is going via Flat DeBruijn.
+FakeNamedDeBruijn<->DeBruijn are isomorphic: we could use iso-deriving package,
+but we do not need any other isomorphic Flat deriving for the moment.
+See NOTE: [Why newtype FakeNamedDeBruijn]
+-}
 instance Flat FakeNamedDeBruijn where
-    size  = size . fromFake -- via debruijn
-    encode  = encode . fromFake -- via debruijn
-    decode =  toFake <$> decode -- via debruijn
-
--- this instance is very similar to the Flat (Binder DeBruijn) instance.
--- NOTE: the serialization roundtrip holds iff the invariant name==fakeName holds
+    size = size . fromFake
+    encode = encode . fromFake
+    decode =  toFake <$> decode
+
+{- This instance is going via Flat (Binder DeBruijn) instance.
+Binder FakeNamedDeBruijn <->Binder DeBruijn are isomorphic:  we could use iso-deriving package,
+but  we do not need any other isomorphic Flat deriving for the moment.
+See NOTE: [Why newtype FakeNamedDeBruijn]
+NOTE: the serialization roundtrip holds iff the invariant binder.index==0 holds
+-}
 instance Flat (Binder FakeNamedDeBruijn) where
-    size  = size . fromFake . coerce -- via binder debruijn
-    encode = encode . fromFake . coerce -- via binder debruijn
-    decode = coerce . toFake . coerce <$> decode @(Binder DeBruijn) -- via binder debruijn
+    size  = size . fmap fromFake
+    encode = encode . fmap fromFake
+    decode =  fmap toFake <$> decode
```

Comment:
Is performance important here? We can do it better performance-wise.

Reply:
I don't think so. But good that u mention it. I wante do to check if there is any ghc-core difference in the decode path.

Reply:
You can probably use `PlutusPrelude.coerceVia (. fmap fromFake) size` and the same for the rest. It's almost as readable, but is definitely faster, although probably negligibly so. But even then I believe it's worth it.

Reply:
I see what you are suggesting, but `Coercible FakeNamedDeBruijn DeBruijn` does not hold. They are isomorphic but they do not have the same runtime representation. So I don't see how I can use the Coercible.coerce mechanism.

Reply:
Oh I see, I just looked at the code and now I understand.

OK, could you please post what GHC Core this instance compiles to?

Reply:
It is very difficult to read the ghc core because I don't get good names for the dictionaries. I was expecting something like $fFlatBinderDeBruijn  and $fFlatBinderFakeNamedDeBruijn but I am getting $fflatBinder + a digit.
this is for all the "Flexible Instances" , e.g. instance Flat (Binder DB) , instance Flat (Binder FND) , instance Flat (Binder NDB).. etc.

But I am fairly sure that the decode method implementation for `Flat (Binder FakeNamedDeBruijn)` is this one, and has remained exactly the same:


```
-- RHS size: {terms: 17, types: 23, coercions: 3, joins: 0/0}
$fFlatBinder7
  = \ _ s s1 ->
      case s of { S dt1 dt2 ->
      (# s1,
         case fakeName of dt3 { Text ipv ipv1 ipv2 ->
         GetResult dt1 dt2 ((NamedDeBruijn dt3 0##) `cast` <Co:3>)
         } #)
      }
```

I will send you the whole diff by email.

Reply:
Yes, decoding only. I dnt know how to improve it. Maybe NOINLINE can help?